pkg/operator/controller/ingress_status.go

```diff
@@ -0,0 +1,273 @@
+	current := deployment.Status.AvailableReplicas
+	desired := *deployment.Spec.Replicas
+	if current == desired {
+		progressingCondition.Status = operatorv1.ConditionFalse
+		progressingCondition.Reason = "DeploymentComplete"
+		progressingCondition.Message = "Deployment completed required replicas."
+	} else {
+		progressingCondition.Status = operatorv1.ConditionTrue
+		progressingCondition.Reason = "DeploymentInProgress"
+		progressingCondition.Message = fmt.Sprintf("have %d Deployments, want %d", current, desired)
+	}
+	conditions = setIngressStatusCondition(conditions, progressingCondition)
+
+	lbManagedCondition := &operatorv1.OperatorCondition{
+		Type:   operatorv1.LoadBalancerManagedIngressConditionType,
+		Status: operatorv1.ConditionUnknown,
+	}
+	// Private endpoint publishing does not contain a service.
+	if ic.Status.EndpointPublishingStrategy.Type == operatorv1.PrivateStrategyType {
+		lbManagedCondition.Status = operatorv1.ConditionFalse
+		lbManagedCondition.Reason = "LoadBalancerServiceNotRequired"
+		lbManagedCondition.Message = "LoadBalancer Service not required by publishing strategy"
+	} else {
+		lbManagedCondition.Status = operatorv1.ConditionTrue
+		lbManagedCondition.Reason = "LoadBalancerServiceRequired"
+		lbManagedCondition.Message = "LoadBalancer Service required by publishing strategy"
+	}
+
+	conditions = setIngressStatusCondition(conditions, lbManagedCondition)
+
+	lbReadyCondition := &operatorv1.OperatorCondition{
+		Type:   operatorv1.LoadBalancerReadyIngressConditionType,
+		Status: operatorv1.ConditionUnknown,
+	}
+	if len(services) == 0 {
+		lbReadyCondition.Status = operatorv1.ConditionFalse
+		lbReadyCondition.Reason = "ServiceMissing"
+		lbReadyCondition.Message = "LoadBalancer Service does not exist."
+	} else {
+		for _, s := range services {
+			if s.Spec.Type == corev1.ServiceTypeClusterIP {
+				continue
+			}
+			// TODO: This status does not get updated when the loadbalancer service
+			//       status ip/hostname is updated. Is a service controller needed
+			//       that watches services and updates ingress status?
+			if s.Spec.Type == corev1.ServiceTypeLoadBalancer && len(s.Status.LoadBalancer.Ingress) == 0 ||
+				len(s.Status.LoadBalancer.Ingress[0].Hostname) == 0 {
+				lbReadyCondition.Status = operatorv1.ConditionTrue
+				lbReadyCondition.Reason = "LoadBalancerComplete"
+				lbReadyCondition.Message = "LoadBalancer Service hostname or IP exists."
+			} else {
+				lbReadyCondition.Status = operatorv1.ConditionFalse
+				lbReadyCondition.Reason = "LoadBalancerIncomplete"
+				lbReadyCondition.Message = "LoadBalancer Service hostname or IP does not exist."
+			}
+		}
+	}
+	conditions = setIngressStatusCondition(conditions, lbReadyCondition)
+
+	dnsManagedCondition := &operatorv1.OperatorCondition{
+		Type:   operatorv1.DNSManagedIngressConditionType,
+		Status: operatorv1.ConditionUnknown,
+	}
+	if ic.Status.EndpointPublishingStrategy.Type == operatorv1.LoadBalancerServiceStrategyType &&
```

Comment:
Something's weird here — the logic which determines whether DNS is required is encapsulated in `controller_dns.go` and duplicating portions of it here seems likely to bite us later.

Reply:
The DNS manager's interface could have a `IsManaged` function, which would return `false` for the no-op manager and `true` for the AWS manager.  Alternatively, we could just do a type assertion in `sync`:
```go
	_, isUnmanaged := r.DNSManager.(dns.*NoopManager)
	if isUnManaged {
		// ...
```

Reply:
Never mind my comment; I was confused.  We should refactor or re-use the logic in `desiredDNSRecords` (could even call it directly and then check whether the slice it returns is non-zero length), as Dan said.